test/unit/feature/isolation.py

```diff
@@ -0,0 +1,78 @@
+import os
+import json
+from unit.applications.proto import TestApplicationProto
+from unit.applications.lang.go import TestApplicationGo
+from unit.applications.lang.java import TestApplicationJava
+from unit.applications.lang.node import TestApplicationNode
+from unit.applications.lang.perl import TestApplicationPerl
+from unit.applications.lang.php import TestApplicationPHP
+from unit.applications.lang.python import TestApplicationPython
+from unit.applications.lang.ruby import TestApplicationRuby
+
+class TestFeatureIsolation(TestApplicationProto):
+    allns = ['pid', 'mnt', 'ipc', 'uts', 'cgroup', 'net']
+
+    def check(self, available, testdir):
+        test_conf = {"namespaces": {"credential": True}}
+
+        module = ''
+        app = 'empty'
+        if 'go' in available['modules']:
+            module = TestApplicationGo()
+
+        elif 'java' in available['modules']:
+            module = TestApplicationJava()
+
+        elif 'node' in available['modules']:
+            module = TestApplicationNode()
+            app = 'basic'
+
+        elif 'perl' in available['modules']:
+            module = TestApplicationPerl()
+            app = 'body_empty'
+
+        elif 'php' in available['modules']:
+            module = TestApplicationPHP()
+            app = 'phpinfo'
+
+        elif 'python' in available['modules']:
+            module = TestApplicationPython()
+
+        elif 'ruby' in available['modules']:
+            module = TestApplicationRuby()
+
+        if not module:
+            return
+
+        module.testdir = testdir
+        module.load(app)
+
+        resp = module.conf(test_conf, 'applications/' + app + '/isolation')
+
+        if 'success' in resp:
+            available['features']['isolation'] = {}
+
+            userns = self.getns('user')
+            if not userns:
+                return
+
+            available['features']['isolation']['user'] = userns
```

Comment:
Probably following diff will be more correct (to avoid set even empty `isolation` field in `available['features']`):
```
diff --git a/test/unit/feature/isolation.py b/test/unit/feature/isolation.py
--- a/test/unit/feature/isolation.py
+++ b/test/unit/feature/isolation.py
@@ -48,15 +48,11 @@ class TestFeatureIsolation(TestApplicati
         module.load(app)
 
         resp = module.conf(test_conf, 'applications/' + app + '/isolation')
+        userns = self.getns('user')
 
-        if 'success' in resp:
-            available['features']['isolation'] = {}
+        if 'success' in resp and userns:
+            available['features']['isolation'] = {'user': userns}
 
-            userns = self.getns('user')
-            if not userns:
-                return
-
-            available['features']['isolation']['user'] = userns
             for ns in self.allns:
                 ns_value = self.getns(ns)
                 if ns_value:

```

Reply:
done.

Have debian tests passed now?

Reply:
> Have debian tests passed now?

No, same alert.

Reply:
Not only debian. Also centos (7.4) and rhel (7.4) - previously tests didn't run on these systems because of hanging in `getns()` call.

Reply:
I did the `/proc/sys/kernel/unprivileged_userns_clone` detection. Then, theoretically, now it should ignore the systems that don't support unprivileged userns.

Thank you

Reply:
Please consider to apply following diff. It removes skip of `test_isolation_values` (checked with buildbot):

```diff --git a/test/test_go_isolation.py b/test/test_go_isolation.py
--- a/test/test_go_isolation.py
+++ b/test/test_go_isolation.py
@@ -18,6 +18,9 @@ class TestGoIsolation(TestApplicationGo)
 
         return unit if not complete_check else unit.complete()
 
+    def isolation_key(self, key):
+        return key in self.available['features']['isolation'].keys()
+
     def conf_isolation(self, isolation):
         self.assertIn(
             'success',
@@ -31,9 +34,16 @@ class TestGoIsolation(TestApplicationGo)
         obj = self.isolation.parsejson(self.get()['body'])
 
         for ns, ns_value in self.available['features']['isolation'].items():
-            self.assertEqual(obj['NS'][ns.upper()], ns_value, '%s match' % ns)
+            if ns.upper() in obj['NS']:
+                self.assertEqual(
+                    obj['NS'][ns.upper()], ns_value, '%s match' % ns
+                )
 
     def test_isolation_user(self):
+        if not self.isolation_key('unprivileged_userns_clone'):
+            print('unprivileged clone is not available')
+            raise unittest.SkipTest()
+
         self.load('ns_inspect')
         obj = self.isolation.parsejson(self.get()['body'])
 
@@ -69,8 +79,12 @@ class TestGoIsolation(TestApplicationGo)
         self.assertEqual(obj['GID'], 1000, 'gid root')
 
     def test_isolation_mnt(self):
-        if 'mnt' not in self.available['features']['isolation'].keys():
-            print('mnt namespace not supported')
+        if not self.isolation_key('mnt'):
+            print('mnt namespace is not supported')
+            raise unittest.SkipTest()
+
+        if not self.isolation_key('unprivileged_userns_clone'):
+            print('unprivileged clone is not available')
             raise unittest.SkipTest()
 
         self.load('ns_inspect')
@@ -86,11 +100,12 @@ class TestGoIsolation(TestApplicationGo)
         allns.remove('mnt')
 
         for ns in allns:
-            self.assertEqual(
-                obj['NS'][ns.upper()],
-                self.available['features']['isolation'][ns],
-                '%s match' % ns,
-            )
+            if ns.upper() in obj['NS']:
+                self.assertEqual(
+                    obj['NS'][ns.upper()],
+                    self.available['features']['isolation'][ns],
+                    '%s match' % ns,
+                )
 
         self.assertNotEqual(
             obj['NS']['MNT'], self.isolation.getns('mnt'), 'mnt set'
@@ -100,8 +115,12 @@ class TestGoIsolation(TestApplicationGo)
         )
 
     def test_isolation_pid(self):
-        if 'pid' not in self.available['features']['isolation'].keys():
-            print('pid namespace not supported')
+        if not self.isolation_key('pid'):
+            print('pid namespace is not supported')
+            raise unittest.SkipTest()
+
+        if not self.isolation_key('unprivileged_userns_clone'):
+            print('unprivileged clone is not available')
             raise unittest.SkipTest()
 
         self.load('ns_inspect')
diff --git a/test/unit/feature/isolation.py b/test/unit/feature/isolation.py
--- a/test/unit/feature/isolation.py
+++ b/test/unit/feature/isolation.py
@@ -9,6 +9,7 @@ from unit.applications.lang.php import T
 from unit.applications.lang.python import TestApplicationPython
 from unit.applications.lang.ruby import TestApplicationRuby
 
+
 class TestFeatureIsolation(TestApplicationProto):
     allns = ['pid', 'mnt', 'ipc', 'uts', 'cgroup', 'net']
 
@@ -48,15 +49,27 @@ class TestFeatureIsolation(TestApplicati
         module.load(app)
 
         resp = module.conf(test_conf, 'applications/' + app + '/isolation')
+        if 'success' not in resp:
+            return
+
         userns = self.getns('user')
+        if not userns:
+            return
 
-        if 'success' in resp and userns and self.have_unpriv_userns():
-            available['features']['isolation'] = {'user': userns}
+        available['features']['isolation'] = {'user': userns}
 
-            for ns in self.allns:
-                ns_value = self.getns(ns)
-                if ns_value:
-                    available['features']['isolation'][ns] = ns_value
+        unp_clone_path = '/proc/sys/kernel/unprivileged_userns_clone'
+        if os.path.exists(unp_clone_path):
+            with open(unp_clone_path, 'r') as f:
+                if str(f.read()).rstrip() == '1':
+                    available['features']['isolation'][
+                        'unprivileged_userns_clone'
+                    ] = True
+
+        for ns in self.allns:
+            ns_value = self.getns(ns)
+            if ns_value:
+                available['features']['isolation'][ns] = ns_value
 
     def getns(self, nstype):
         # read namespace id from symlink file:
@@ -70,17 +83,5 @@ class TestFeatureIsolation(TestApplicati
 
         return data
 
-    def have_unpriv_userns(self):
-        userns_config = "/proc/sys/kernel/unprivileged_userns_clone"
-        if os.path.exists(userns_config):
-            f = open(userns_config, "r")
-            val = f.read()
-            f.close()
-            if str(val).rstrip() == "1":
-                return True
-
-        return False
-
-
     def parsejson(self, data):
         return json.loads(data.split('\n')[1])
```

Reply:
Thanks

Reply:
Thank you!